crates/compilers/src/preprocessor.rs

```diff
@@ -0,0 +1,643 @@
+                        // remove only the body.
+                        if !is_exposed {
+                            spans_to_remove.push(contract_item.span);
+                        } else {
+                            spans_to_remove.push(function.body_span);
+                        }
+                    }
+                }
+            }
+        }
+    });
+
+    // Return if any diagnostics emitted during content parsing.
+    if let Err(err) = sess.emitted_errors().unwrap() {
+        let e = err.to_string();
+        trace!("failed parsing {file:?}: {e}");
+        return Err(err);
+    }
+
+    let mut content = content.to_string();
+    let mut offset = 0;
+
+    for span in spans_to_remove {
+        let range = span.to_range();
+        let start = range.start - offset;
+        let end = range.end - offset;
+
+        content.replace_range(start..end, "");
+        offset += end - start;
+    }
+
+    let content = content.replace("\n", "");
+    Ok(utils::RE_TWO_OR_MORE_SPACES.replace_all(&content, "").to_string())
+}
+
+/// Computes hash of [`interface_representation`] of the source.
+pub(crate) fn interface_representation_hash(source: &Source, file: &PathBuf) -> String {
+    let Ok(repr) = interface_representation(&source.content, file) else {
+        return source.content_hash();
+    };
+    let mut hasher = md5::Md5::new();
+    hasher.update(&repr);
+    let result = hasher.finalize();
+    hex::encode(result)
+}
+
+#[derive(Debug)]
+pub struct ItemLocation {
+    start: usize,
+    end: usize,
+}
+
+impl ItemLocation {
+    fn try_from_loc(loc: SourceLocation) -> Option<Self> {
+        Some(Self { start: loc.start?, end: loc.start? + loc.length? })
+    }
+}
+
+/// Checks if the given path is a test/script file.
+fn is_test_or_script<L>(path: &Path, paths: &ProjectPathsConfig<L>) -> bool {
+    let test_dir = paths.tests.strip_prefix(&paths.root).unwrap_or(&paths.root);
+    let script_dir = paths.scripts.strip_prefix(&paths.root).unwrap_or(&paths.root);
+    path.starts_with(test_dir) || path.starts_with(script_dir)
+}
+
+/// Kind of the bytecode dependency.
+#[derive(Debug)]
+enum BytecodeDependencyKind {
+    /// `type(Contract).creationCode`
+    CreationCode,
+    /// `new Contract`
+    New(ItemLocation, String),
+}
+
+/// Represents a single bytecode dependency.
+#[derive(Debug)]
+struct BytecodeDependency {
+    kind: BytecodeDependencyKind,
+    loc: ItemLocation,
+    referenced_contract: usize,
+}
+
+/// Walks over AST and collects [`BytecodeDependency`]s.
+#[derive(Debug)]
+struct BytecodeDependencyCollector<'a> {
+    source: &'a str,
+    dependencies: Vec<BytecodeDependency>,
+    total_count: usize,
+}
+
+impl BytecodeDependencyCollector<'_> {
+    fn new(source: &str) -> BytecodeDependencyCollector<'_> {
+        BytecodeDependencyCollector { source, dependencies: Vec::new(), total_count: 0 }
+    }
+}
+
+impl Visitor for BytecodeDependencyCollector<'_> {
```

Comment:
this should use solar HIR, requires patching to latest git

Reply:
@DaniPopes I am having some difficulties figuring out how to use HIR API (trying smth like below), could you please shed some light? thank you!

```rust
        sess.enter_parallel(|| -> solar_parse::interface::Result {
            let mut parsing_context = ParsingContext::new(&sess);
            let arena: ThreadLocal<Arena> = ThreadLocal::new();
            parsing_context.load_files(input.input.sources.iter().map(|(path, _)| path.as_path()))?;
            let parsed_sources = parsing_context.parse(&arena);
            /// Do something with parsed sources
            Ok(())
        });
```

Reply:
@DaniPopes got an working impl, please recheck :pray: